src/rust/src/encoder/g608.rs

```diff
@@ -0,0 +1,308 @@
+use crate::bindings::{
+    ccx_encoding_type_CCX_ENC_ASCII, ccx_encoding_type_CCX_ENC_LATIN_1,
+    ccx_encoding_type_CCX_ENC_UNICODE, ccx_encoding_type_CCX_ENC_UTF_8, eia608_screen, encoder_ctx,
+    font_bits_FONT_ITALICS, font_bits_FONT_REGULAR, font_bits_FONT_UNDERLINED,
+    font_bits_FONT_UNDERLINED_ITALICS,
+};
+use crate::encoder::common::{encode_line, write_raw};
+use crate::libccxr_exports::time::ccxr_millis_to_time;
+use lib_ccxr::util::encoding::{line21_to_latin1, line21_to_ucs2, line21_to_utf8};
+use std::io;
+use std::os::raw::{c_int, c_uchar, c_uint, c_void};
+
+/// Write data to file descriptor with retry logic
+pub fn write_wrapped(fd: c_int, buf: &[u8]) -> Result<(), io::Error> {
+    let mut remaining = buf.len();
+    let mut current_buf = buf.as_ptr();
+
+    while remaining > 0 {
+        let written = write_raw(fd, current_buf as *const c_void, remaining);
+        if written == -1 {
+            return Err(io::Error::last_os_error());
+        }
+        let bytes_written = written as usize;
+        unsafe {
+            current_buf = current_buf.add(bytes_written);
+        }
+        remaining -= bytes_written;
+    }
+
+    Ok(())
+}
+
+/// Get line characters encoded according to context encoding
+pub fn get_line_encoded(
+    ctx: &encoder_ctx,
+    buffer: &mut [c_uchar],
+    line_num: usize,
+    data: &eia608_screen,
+) -> c_uint {
+    let mut buffer_pos = 0;
+    let line = &data.characters[line_num];
+
+    for i in line.iter().take(33) {
+        let bytes_written = match ctx.encoding {
+            ccx_encoding_type_CCX_ENC_UTF_8 => {
+                let utf8_packed = line21_to_utf8(*i);
+                let (byte_count, bytes) = if utf8_packed <= 0xff {
+                    (1, [(utf8_packed & 0xFF) as c_uchar, 0, 0])
+                } else if utf8_packed <= 0xffff {
+                    (
+                        2,
+                        [
+                            ((utf8_packed >> 8) & 0xFF) as c_uchar,
+                            (utf8_packed & 0xFF) as c_uchar,
+                            0,
+                        ],
+                    )
+                } else {
+                    (
+                        3,
+                        [
+                            ((utf8_packed >> 16) & 0xFF) as c_uchar,
+                            ((utf8_packed >> 8) & 0xFF) as c_uchar,
+                            (utf8_packed & 0xFF) as c_uchar,
+                        ],
+                    )
+                };
+
+                if buffer_pos + byte_count - 1 < buffer.len() {
+                    buffer[buffer_pos..(byte_count + buffer_pos)]
```

Comment:
This should be like this after you modify line21_to_utf8 to return vector of u8's, I have DM'ed you with more details:
```suggestion
                let bytes = line21_to_utf8(*i);
                byte_count = bytes.len();

                buffer[buffer_pos] = bytes[0] as c_uchar;
                buffer[buffer_pos + 1] = bytes[1] as c_uchar;
                buffer[buffer_pos + 2] = bytes[2] as c_uchar;

                byte_count
```